# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: testing/web-platform/tests/mediacapture-record/MediaRecorder-destroy-script-execution.html
# Commit: 3f2015f34fc2
# Full Hash: 3f2015f34fc2daaaff9d90b7114157face56edf2
# Author: Olli Pettay <Olli.Pettay@helsinki.fi>
# Date: 2022-09-14 04:09:22
# Description:
#   Bug 1713940, null check owner before using it, r=padenot
#   
#   Reused an existing wpt. Added also some .step() calls there.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D157143
# ==============================================================================

diff -r dff049971660 -r 3f2015f34fc2 testing/web-platform/tests/mediacapture-record/MediaRecorder-destroy-script-execution.html
--- a/testing/web-platform/tests/mediacapture-record/MediaRecorder-destroy-script-execution.html	Tue Sep 13 17:08:46 2022 +0000
+++ b/testing/web-platform/tests/mediacapture-record/MediaRecorder-destroy-script-execution.html	Tue Sep 13 17:09:24 2022 +0000
@@ -5,18 +5,35 @@
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <body>
+<iframe src="support/MediaRecorder-iframe.html" id="subFrame-start" name="subFrameStart"></iframe>
 <iframe src="support/MediaRecorder-iframe.html" id="subFrame-stop" name="subFrameStop"></iframe>
 <iframe src="support/MediaRecorder-iframe.html" id="subFrame-allTrackEnded" name="subFrameAllTrackEnded"></iframe>
 <iframe src="support/MediaRecorder-iframe.html" id="subFrame-audioBitrateMode" name="subFrameAudioBitrateMode"></iframe>
 <script>
+    var iframeForCallingStart = document.getElementById('subFrame-start');
     var iframeForCallingStop = document.getElementById('subFrame-stop');
     var iframeForAllTrackEnded = document.getElementById('subFrame-allTrackEnded');
     var iframeForAudioBitrateMode = document.getElementById('subFrame-audioBitrateMode');
 
+    var testForCallingStart = async_test('MediaRecorder will throw when start() is called and the script execution context is going away');
     var testForCallingStop = async_test('MediaRecorder will not fire the stop event when stop() is called and the script execution context is going away');
     var testForAllTrackEnded = async_test('MediaRecorder will not fire the stop event when all tracks are ended and the script execution context is going away');
     var testForAudioBitrateMode = async_test('MediaRecorder will not crash on accessing audioBitrateMode when the script execution context is going away');
 
+
+    iframeForCallingStart.onload = function(e) {
+      let testWindow = subFrameStart.window;
+      testWindow.prepareForTest(testForCallingStart);
+      let exceptionCtor = testWindow.DOMException;
+      const recorder = subFrameStart.window.recorder;
+      iframeForCallingStart.remove();
+      testForCallingStart.step(function() {
+        assert_throws_dom('NotSupportedError', exceptionCtor, () => recorder.start(),
+                         "MediaRecorder.start() should throw");
+      });;
+      testForCallingStart.done();
+    };
+
     iframeForCallingStop.onload = function(e) {
         subFrameStop.window.prepareForTest(testForCallingStop);
         const recorder = subFrameStop.window.recorder;
@@ -26,7 +43,9 @@
         });
         recorder.onstop = testForCallingStop.unreached_func('Unexpected stop event');
         recorder.start();
-        assert_equals(recorder.state, 'recording', 'MediaRecorder has been started successfully');
+        testForCallingStop.step(function() {
+          assert_equals(recorder.state, 'recording', 'MediaRecorder has been started successfully');
+        });
         subFrameStop.window.control.addVideoFrame();
         recorder.stop();
     };
@@ -40,7 +59,9 @@
         });
         recorder.onstop = testForAllTrackEnded.unreached_func('Unexpected stop event');
         recorder.start();
-        assert_equals(recorder.state, 'recording', 'MediaRecorder has been started successfully');
+        testForAllTrackEnded.step(function() {
+          assert_equals(recorder.state, 'recording', 'MediaRecorder has been started successfully');
+        });
         subFrameAllTrackEnded.window.control.addVideoFrame();
         subFrameAllTrackEnded.window.video.getVideoTracks()[0].stop();
     };
